src/tunnel.cc

```diff
@@ -323,30 +323,35 @@ TunnelStateData::handleClientClosure()
         server.conn->close();
 }
 
-/// Server close handler applied on the TLS connection being established.
-/// If TLS handshake failures, retries connection to another destination.
+/// handles the closure of a being-established Squid-to-server TLS connection
 static void
 tlsServerClosed(const CommCloseCbParams &params)
 {
     const auto tunnelState = reinterpret_cast<TunnelStateData *>(params.data);
     debugs(26, 3, tunnelState->server.conn);
     tunnelState->server.resetCloseHandler();
+
     if (!tunnelState->destinations->empty()) {
         if (!tunnelState->opening())
             tunnelState->startConnecting(); // try connecting to another destination
-    } else if (tunnelState->subscribed) {
-        if (!tunnelState->hasError()) {
-            const auto anErr = new ErrorState(ERR_CANNOT_FORWARD, Http::scInternalServerError,
-                    tunnelState->request.getRaw(), tunnelState->al);
-            tunnelState->saveError(anErr);
-        } // else use actual error from last connection attempt
+        return;
+    }
+
+    if (tunnelState->subscribed) {
         debugs(26, 4, "wait for more destinations to try");
-    } else {
-        tunnelState->handleServerClosure();
+        return;
     }
+
+    if (!tunnelState->hasError()) {
+        const auto anErr = new ErrorState(ERR_CANNOT_FORWARD, Http::scInternalServerError,
+                tunnelState->request.getRaw(), tunnelState->al);
+        tunnelState->saveError(anErr);
+    } // else use actual error from last connection attempt
+
+    tunnelState->handleServerClosure();
```

Comment:
> We
need to stop only when there are no destinations to try _and_ no more
destinations are expected

The difference I see between the old code and the new one is where we check and save the error by saveError(): previously it was done if subscribed is true, now - if subscribed is false. If my understanding is correct, the old code could work incorrectly because we saved the error before all destinations have been tried and noteDestinationsEnd() was called: during checking the last destination noteDestinationsEnd() could see the saved error and do preliminary abort by sendError(). However, the same problem can happen because we also save the error in connectedToPeer(), i.e., before tlsServerClosed().

Reply:
> the old code could work incorrectly because we saved the error before all destinations have been tried

AFAICT, the old code was wrong because it did premature error handling/processing (i.e. checking whether there was a saved error and creating a new made-up error if there was not) and initiated tunnel termination (i.e. calling handleServerClosure()) instead of simply waiting for more destinations to try. In other words, the old code treated the tunnel as unusable/dead/broken when there was no reason for doing that.

> during checking the last destination noteDestinationsEnd() could see the saved error

FWIW, I do not think that specific series of events was possible -- noteDestinationsEnd() does not check for saved errors if destinationsFound is true. The noteDestination() method naturally makes destinationsFound true. And without at least one noteDestination() call, there can be no tlsServerClosed(), right?

Reply:
> (i.e. checking whether there was a saved error and creating a new made-up error

Agreed. I thought that we need to save the error from the previous (unsuccessful) attempt so that we could use it if something went wrong (and no more errors are available). Probably this situation is unlikely, and each connection attempt is either successful or produces an error.

> and initiated tunnel termination (i.e. calling handleServerClosure()) instead of simply waiting for more destinations to try.

Disagreed, this is not how the old code worked. Basically, you changed my if/elses into if/returns, but the logic remains the same.

Before:

```
 if (!tunnelState->destinations->empty()) {
    // start connecting
 } else if (tunnelState->subscribed) {
   // create an error if needed and wait
 } else {
  // cleanup
 }
```

After:

```
if (!tunnelState->destinations->empty()) {
  // start connecting
  return;
}
if (tunnelState->subscribed) {
  // wait
  return;
}
// save the error if needed
// cleanup

```

> And without at least one noteDestination() call, there can be no tlsServerClosed(), right?

I think you are right.

Reply:
> I thought that we need to save the error from the previous (unsuccessful) attempt so that we could use it if something went wrong (and no more errors are available).

You thought correctly, but connection closure itself is not an error. If we discovered an error earlier, then it was already saved. Otherwise, (i.e. this is some kind of external closure that we did not initiate), there is no information we _can_ save to explain the being-closed connection -- we do not know why it got closed or who closed it. All we can do is to try the next destination.

If there are no other destinations to try (or retries are prohibited), then we reuse the previously saved error or (if there was not one saved), generate a generic "we are out of paths" error.

If you are OK with the above explanation and the current tlsServerClosed() code, then I believe this conversation thread can be resolved because the code already reflects what was said here.

----

> Disagreed, this is not how the old code worked

You are right. I probably misplaced the handleServerClosure() call when describing the diff. Sorry. Only the error creation was wrong, not the handleServerClosure() call placement.